util/gatewaycrafter/gatewaycrafter.go

```diff
@@ -0,0 +1,86 @@
+package gatewaycrafter
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/earthly/earthly/states/image"
+
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/pkg/errors"
+)
+
+// GatewayCrafter abstracts the logic required to interact with the gateway Build and Export functions
+type GatewayCrafter interface {
+	AddPushImageEntry(ref gwclient.Reference, refID int, imageName string, shouldPush, insecurePush bool, imageConfig *image.Image, platformStr []byte) (string, error)
+}
+
+// MapBasedGatewayCrafter extends the GatewayCrafter but provides an additional method for
+// fetching the Ref and metadata maps which must be passed to the gwclient.Export method.
+type MapBasedGatewayCrafter interface {
+	GatewayCrafter
+	GetRefsAndMetadata() (map[string]gwclient.Reference, map[string][]byte)
+}
```

Comment:
Feels like this could be simplified.

Instead of defining `gatewayCrafterBase` as a set of functions, we could simply make it

```go
type GatewayCrafeter struct {
  refs map[string]gwclient.Reference{}
  meta map[string][]byte
}
```

Then there would only have to be a single implementation (no interfaces). It just that the caller decides to either extract the information contained within as a `*gwclient.Result` or as refs+meta.

```go
func (c *GatewayCrafter) AsGatewayClientResult() (*gwclient.Result) {
  res := gwclient.NewResult()
  for k, v := range c.refs {
     res.AddRef(k, v)
  }
  for k, v := range c.meta {
     res.AddMeta(k, v)
  }
  return res
}
```

```go
func (c *GatewayCrafter) AsRefsAndMeta() (...) {
  return c.refs, c.meta
}
```

The constructor can then be simply

```go
func NewGatewayCrafter() *GatewayCrafter {
  return &GatewayCrafter{
    meta: make(...),
    refs: make(...),
  }
}
```

Reply:
I don't think this will work because the builder/builder.go code still requires that part of the results object be modified directly; we need to either pass in a reference (like we already do), or simply make a call to `AsRefsAndMeta()`, then merge that data into the existing results.

For example here's the builder code:

```
		res := gwclient.NewResult()
		gwCrafter := gatewaycrafter.NewGatewayCrafterForBuild(res)
		if !b.builtMain {
			ref, err := b.stateToRef(childCtx, gwClient, mts.Final.MainState, mts.Final.PlatformResolver)
			if err != nil {
				return nil, err
			}
			res.AddRef("main", ref) <--- This adds directly to the results
		}
```

the part way down the function, we have:

```
					// For push.
					if shouldPush {
						_, err = gwCrafter.AddPushImageEntry(ref, imageIndex, saveImage.DockerTag, shouldPush, saveImage.InsecurePush, saveImage.Image, []byte(platformStr))
						if err != nil {
							return nil, err
						}
						imageIndex++
					}
```

we could change the function to simply export `AsRefsAndMeta()`, where we then iterate over the refs and metadata and merge it back into res, e.g.

```
res := gwclient.NewResult()

for .... {
... code that works directly on res
....and code that calls gwCrafter.AddPushImageEntry(....)
}

...
// somewhere after the loop

refsToMerge, metadataToMerge := gwCrafter.AsRefsAndMeta()
for k, v := range refsToMerge {
    res.AddRef(k, v)
}
for k, v := range metadataToMerge {
    res.AddMeta(k, v)
}
```

Reply:
We could perhaps expose `AddRef` in the crafter for that situation?

Reply:
Maybe both `AddRef` and `AddMeta`, at least until we port all types of exports for the crafter to handle elegantly.

Reply:
That could work!